### Import our libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline 

### Read our csv files into a dataframe

In [2]:
rawCallData = pd.read_csv('Data/Seattle_Police_Department_911_Incident_Response.csv')

/Users/BenM/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### An error right out of the gate, let see what the problem is

In [3]:
rawCallData.dtypes

CAD CDW ID                      object
CAD Event Number                 int64
General Offense Number           int64
Event Clearance Code           float64
Event Clearance Description     object
Event Clearance SubGroup        object
Event Clearance Group           object
Event Clearance Date            object
Hundred Block Location          object
District/Sector                 object
Zone/Beat                       object
Census Tract                   float64
Longitude                      float64
Latitude                       float64
Incident Location               object
Initial Type Description        object
Initial Type Subgroup           object
Initial Type Group              object
At Scene Time                   object
dtype: object

### Let's start by cleaning up those column names in rawCallData

In [4]:
rawCallData.columns = rawCallData.columns.str.lower().str.replace(" ", "_").str.replace("/", "_") 
# Remove the white space and slashes in our column names
rawCallData.columns # Check our work

Index(['cad_cdw_id', 'cad_event_number', 'general_offense_number',
       'event_clearance_code', 'event_clearance_description',
       'event_clearance_subgroup', 'event_clearance_group',
       'event_clearance_date', 'hundred_block_location', 'district_sector',
       'zone_beat', 'census_tract', 'longitude', 'latitude',
       'incident_location', 'initial_type_description',
       'initial_type_subgroup', 'initial_type_group', 'at_scene_time'],
      dtype='object')

### That's better,  but an attempt to coerce the datatype leads us to the problems,
Let's take a minute and fix these

In [5]:
rawCallData.loc[rawCallData.cad_cdw_id == '1875911']

cad_cdw_id  cad_event_number  general_offense_number  \
857606  1875911       14000012658               201412658   

        event_clearance_code event_clearance_description  \
857606                  41.0         HARASSMENT, THREATS   

       event_clearance_subgroup event_clearance_group    event_clearance_date  \
857606      THREATS, HARASSMENT   THREATS, HARASSMENT  01/13/2014 12:51:00 AM   

           hundred_block_location district_sector zone_beat  census_tract  \
857606  39XX BLOCK OF S JUNEAU ST               S        S1    10300.3006   

         longitude   latitude               incident_location  \
857606 -122.282103  47.549712  (47.549711879, -122.282103194)   

                             initial_type_description initial_type_subgroup  \
857606  THREATS (INCLS IN-PERSON/BY PHONE/IN WRITING)   THREATS, HARASSMENT   

         initial_type_group           at_scene_time  
857606  THREATS, HARASSMENT  01/12/2014 08:00:00 PM

In [6]:
rawCallData.loc[857606,'cad_cdw_id'] = '1875911'

In [7]:
rawCallData.loc[rawCallData.cad_cdw_id == '\ufeff15736']

cad_cdw_id  cad_event_number  general_offense_number  event_clearance_code  \
0     15736       10000246357              2010246357                 242.0   

  event_clearance_description event_clearance_subgroup event_clearance_group  \
0           FIGHT DISTURBANCE             DISTURBANCES          DISTURBANCES   

     event_clearance_date hundred_block_location district_sector zone_beat  \
0  07/17/2010 08:49:00 PM   3XX BLOCK OF PINE ST               M        M2   

   census_tract   longitude   latitude               incident_location  \
0     8100.2001 -122.338147  47.610975  (47.610975163, -122.338146748)   

  initial_type_description initial_type_subgroup initial_type_group  \
0                      NaN                   NaN                NaN   

  at_scene_time  
0           NaN

In [8]:
rawCallData.loc[0, 'cad_cdw_id'] = '15736'

In [9]:
rawCallData.loc[rawCallData.cad_cdw_id == '880295']

cad_cdw_id  cad_event_number  general_offense_number  \
926010  880295       12000276191              2012276191   

        event_clearance_code event_clearance_description  \
926010                  71.0                  AUTO THEFT   

       event_clearance_subgroup event_clearance_group    event_clearance_date  \
926010              AUTO THEFTS           AUTO THEFTS  08/19/2012 03:16:00 PM   

              hundred_block_location district_sector zone_beat  census_tract  \
926010  18XX BLOCK OF EASTLAKE AVE E               D        D2        6600.2   

         longitude  latitude               incident_location  \
926010 -122.325401  47.63537  (47.635369535, -122.325400816)   

                          initial_type_description initial_type_subgroup  \
926010  AUTO THEFT - VEH THEFT OR THEFT & RECOVERY           AUTO THEFTS   

       initial_type_group           at_scene_time  
926010    AUTO RECOVERIES  08/19/2012 02:38:00 PM

In [10]:
rawCallData.loc[926010, 'cad_cdw_id'] = '880295'

In [11]:
rawCallData.cad_cdw_id = rawCallData.cad_cdw_id.astype('int64', errors='raise') # Coerce the data to int64

### Fixed that, on we go to our next dataframe

In [12]:
rawWeatherData = pd.read_csv('Data/Seattle_Weather.csv')

### Lets take a look at our data frames to see what we've got 

In [13]:
rawCallData.head()

cad_cdw_id  cad_event_number  general_offense_number  event_clearance_code  \
0       15736       10000246357              2010246357                 242.0   
1       15737       10000246471              2010246471                  65.0   
2       15738       10000246255              2010246255                 250.0   
3       15739       10000246473              2010246473                 460.0   
4       15740       10000246330              2010246330                 250.0   

     event_clearance_description       event_clearance_subgroup  \
0              FIGHT DISTURBANCE                   DISTURBANCES   
1          THEFT - MISCELLANEOUS                          THEFT   
2  MISCHIEF, NUISANCE COMPLAINTS  NUISANCE, MISCHIEF COMPLAINTS   
3     TRAFFIC (MOVING) VIOLATION          TRAFFIC RELATED CALLS   
4  MISCHIEF, NUISANCE COMPLAINTS  NUISANCE, MISCHIEF COMPLAINTS   

   event_clearance_group    event_clearance_date  \
0           DISTURBANCES  07/17/2010 08:49:00 PM   
1         OTHER PROPERTY  07/17/2010 08:50:00 PM   
2    NUISANCE, MISCHIEF   07/17/2010 08:55:00 PM   
3  TRAFFIC RELATED CALLS  07/17/2010 09:00:00 PM   
4    NUISANCE, MISCHIEF   07/17/2010 09:00:00 PM   

              hundred_block_location district_sector zone_beat  census_tract  \
0               3XX BLOCK OF PINE ST               M        M2     8100.2001   
1  36XX BLOCK OF DISCOVERY PARK BLVD               Q        Q1     5700.1012   
2              21XX BLOCK OF 3RD AVE               M        M2     7200.2025   
3                7XX BLOCK OF ROY ST               D        D1     7200.1002   
4              9XX BLOCK OF ALOHA ST               D        D1     6700.1009   

    longitude   latitude               incident_location  \
0 -122.338147  47.610975  (47.610975163, -122.338146748)   
1 -122.404613  47.658325  (47.658324899, -122.404612874)   
2 -122.342843  47.613551  (47.613551471, -122.342843234)   
3 -122.341847  47.625401  (47.625401388, -122.341846999)   
4 -122.339709  47.627425  (47.627424837, -122.339708605)   

  initial_type_description initial_type_subgroup initial_type_group  \
0                      NaN                   NaN                NaN   
1                      NaN                   NaN                NaN   
2                      NaN                   NaN                NaN   
3                      NaN                   NaN                NaN   
4                      NaN                   NaN                NaN   

  at_scene_time  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

In [14]:
rawWeatherData.head()

dt                         dt_iso  city_id  city_name  lat  lon  \
0  1349096400  2012-10-01 13:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
1  1349186400  2012-10-02 14:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
2  1349190000  2012-10-02 15:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
3  1349193600  2012-10-02 16:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
4  1349197200  2012-10-02 17:00:00 +0000 UTC  5809844        NaN  NaN  NaN   

     temp  temp_min  temp_max  pressure      ...       rain_today  snow_1h  \
0  281.80    278.15    287.59      1027      ...              NaN      NaN   
1  281.62    278.15    286.48      1046      ...              NaN      NaN   
2  282.71    279.82    289.82      1026      ...              NaN      NaN   
3  285.05    281.48    293.15      1026      ...              NaN      NaN   
4  287.97    282.59    296.48      1027      ...              NaN      NaN   

   snow_3h  snow_24h  snow_today  clouds_all  weather_id  weather_main  \
0      NaN       NaN         NaN           1         800         Clear   
1      NaN       NaN         NaN          66         800         Clear   
2      NaN       NaN         NaN           1         800         Clear   
3      NaN       NaN         NaN           1         800         Clear   
4      NaN       NaN         NaN           1         800         Clear   

   weather_description  weather_icon  
0         sky is clear           01n  
1         sky is clear           02d  
2         sky is clear           01d  
3         sky is clear           01d  
4         sky is clear           01d  

[5 rows x 28 columns]

### Lets examine the NaN values, starting with rawCallData

In [16]:
print(rawCallData.isnull().sum())

cad_cdw_id                           0
cad_event_number                     0
general_offense_number               0
event_clearance_code             10797
event_clearance_description      10798
event_clearance_subgroup         10798
event_clearance_group            10798
event_clearance_date             10951
hundred_block_location            3487
district_sector                   1162
zone_beat                            1
census_tract                      2792
longitude                            1
latitude                             1
incident_location                    1
initial_type_description        577813
initial_type_subgroup           577813
initial_type_group              577813
at_scene_time                  1029344
dtype: int64


### We have a significant number of records missing some information
They seem to be clustered. There is a cluster of 10,790+ records missing event data, 577,813 records missing initial data and a third large cluster where most of the records are missing  at scene time. The event clearance and initial type columns seem to all describe the same data, what the call was about. The at scene time and call clearance date similiarly overlap. The bad news is, most of the missing data pertains to what we care about, which is what happened and when. They good news is these columns seem to provide redundant information, so we can use one to impute the other. We can see a few columns provide no relevant information and can be dropped, subgroup and group. Let's see if there are any records that provide no relevant information.

In [17]:
rawCallData.drop(['event_clearance_subgroup', 'event_clearance_group', 'initial_type_subgroup', 
                  'initial_type_group'], axis=1,inplace=True)

In [18]:
mask = (rawCallData.event_clearance_description.isnull()) & (rawCallData.initial_type_description.isnull())
noEvent = rawCallData[mask] # Our mask selects records that have a null value in both description columns
print(noEvent.shape) # Check the size of our haul

(932, 15)


### We have 932 records with no event descriptor, we will have to remove them,
While not shown here we also checked against the other descriptive columns

In [19]:
print(rawCallData.shape) # Check original datafile shape
rawCallData = rawCallData[~mask] # Remove by selecting the inverse of our mask as subset
print(rawCallData.shape) # Verify our subtraction

(1445066, 15)
(1444134, 15)


In [20]:
mask2 = (rawCallData.at_scene_time.isnull()) & (rawCallData.event_clearance_date.isnull())
noTime = rawCallData[mask2] # Our mask selects records that have a null value in both time columns
print(noTime.shape) # Check the size of our haul

(9, 15)


### We have 9 records with no time, we will have to remove them

In [21]:
print(rawCallData.shape) # Check original datafile shape
rawCallData = rawCallData[~mask2] # Remove by selecting the inverse of our mask as subset
print(rawCallData.shape) # Verify our subtraction

(1444134, 15)
(1444125, 15)


### Looking again at the time values,
We can only use the overlap in our data sets date range, so lets switch gears and make sure that our dataframes cover the same time period. Continuing to clean up the null values in rawCallData could be a waste if those records don't overlap with our weather data. We will start by cleaning up the assorted time columns and getting one formatted datetime column.

In [22]:
rawCallData['formatted_time'] = pd.to_datetime(rawCallData.event_clearance_date, 
                                               format='%m/%d/%Y %I:%M:%S %p', errors='coerce')

In [23]:
rawCallData.formatted_time.head()

0   2010-07-17 20:49:00
1   2010-07-17 20:50:00
2   2010-07-17 20:55:00
3   2010-07-17 21:00:00
4   2010-07-17 21:00:00
Name: formatted_time, dtype: datetime64[ns]

In [24]:
rawWeatherData['formatted_time'] = pd.to_datetime(rawWeatherData.dt, unit='s')

In [25]:
print(rawWeatherData.formatted_time.head())
print(rawWeatherData.head())

0   2012-10-01 13:00:00
1   2012-10-02 14:00:00
2   2012-10-02 15:00:00
3   2012-10-02 16:00:00
4   2012-10-02 17:00:00
Name: formatted_time, dtype: datetime64[ns]
           dt                         dt_iso  city_id  city_name  lat  lon  \
0  1349096400  2012-10-01 13:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
1  1349186400  2012-10-02 14:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
2  1349190000  2012-10-02 15:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
3  1349193600  2012-10-02 16:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
4  1349197200  2012-10-02 17:00:00 +0000 UTC  5809844        NaN  NaN  NaN   

     temp  temp_min  temp_max  pressure         ...          snow_1h  snow_3h  \
0  281.80    278.15    287.59      1027         ...              NaN      NaN   
1  281.62    278.15    286.48      1046         ...              NaN      NaN   
2  282.71    279.82    289.82      1026         ...              NaN      NaN   
3  285.05    281.48    293.15      1026    

### We can see this matches the UTC time given in the next column, so we need to convert to PDT.
This is a three step process, first changing our naive time to UTC zone aware time, then converting to PDT through a lambda function. We can't use tz_convert inline because it defaults to changing the index rather than the value. We then undo making it zone aware with tz_localize as another lambda function.

In [26]:
rawWeatherData.formatted_time = rawWeatherData.formatted_time.dt.tz_localize('UTC')
rawWeatherData.formatted_time = rawWeatherData.formatted_time.apply(lambda x: x.tz_convert('America/Los_Angeles'))
rawWeatherData.formatted_time = rawWeatherData.formatted_time.apply(lambda x: x.tz_localize(None))

rawWeatherData.formatted_time.head()

0   2012-10-01 06:00:00
1   2012-10-02 07:00:00
2   2012-10-02 08:00:00
3   2012-10-02 09:00:00
4   2012-10-02 10:00:00
Name: formatted_time, dtype: datetime64[ns]

In [27]:
mask3 = (rawCallData.formatted_time > '2012-10-01 13:00:00')
rawCallData = rawCallData[mask3]

In [28]:
rawCallData.shape

(879393, 16)

In [29]:
rawCallData.isnull().sum()

cad_cdw_id                          0
cad_event_number                    0
general_offense_number              0
event_clearance_code                0
event_clearance_description         0
event_clearance_date                0
hundred_block_location              1
district_sector                   937
zone_beat                           0
census_tract                     1615
longitude                           0
latitude                            0
incident_location                   0
initial_type_description        72663
at_scene_time                  516351
formatted_time                      0
dtype: int64

# This is an important learning note:
# _Always start with a plan_
To be honest when cleaning this data, I just started with the 911 call data, looking for ways to clean out the NaN values. I'm leaving the original plan in this notebook, which was to use one column to impute the other, without regard to the bigger picture. This involved some neat data cleaning tricks, but in the end would have been completely wasted work. Without the corresponding weather data, those records are worthless to us. Luckily, I stopped and thought it out after hitting a few road blocks, realizing that the work might not be necessary. Turns out had I stopped and made a plan in the first place, I could've save quite a bit of time.
## _Always make a plan and then tackle your data, don't just start coding_
### Duly noted, on we go

### The columns with NaN values still are offering us redundant information about the description or location and can be dropped

In [30]:
rawCallData.drop(['event_clearance_date', 'hundred_block_location', 'district_sector', 'census_tract',
                 'incident_location', 'initial_type_description', 'at_scene_time'], inplace=True, axis=1)
rawCallData.shape

(879393, 9)

In [31]:
print(rawCallData.isnull().sum())
rawCallData.head()

cad_cdw_id                     0
cad_event_number               0
general_offense_number         0
event_clearance_code           0
event_clearance_description    0
zone_beat                      0
longitude                      0
latitude                       0
formatted_time                 0
dtype: int64


cad_cdw_id  cad_event_number  general_offense_number  \
49      1658027       16000028163               201628163   
70      1658028       16000028161               201628161   
105     1658029       16000028159               201628159   
190     1658030       16000028134               201628134   
255     1658031       16000028114               201628114   

     event_clearance_code           event_clearance_description zone_beat  \
49                  245.0                    DISTURBANCE, OTHER        N2   
70                  280.0                     SUSPICIOUS PERSON        S1   
105                  65.0                 THEFT - MISCELLANEOUS        M1   
190                 200.0  ALACAD - COMMERCIAL BURGLARY (FALSE)        U1   
255                 161.0                              TRESPASS        D2   

      longitude   latitude      formatted_time  
49  -122.347770  47.731678 2016-01-24 11:54:55  
70  -122.280685  47.523026 2016-01-24 11:57:35  
105 -122.342000  47.609535 2016-01-24 11:54:28  
190 -122.313020  47.668995 2016-01-24 11:53:22  
255 -122.344670  47.641580 2016-01-24 11:59:45

### Now we can look to see if the event numbers and ID numbers offer unique ID's

In [32]:
print(rawCallData.cad_cdw_id.nunique())
print(rawCallData.cad_event_number.nunique())
print(rawCallData.general_offense_number.nunique())

879366
878480
878480


In [33]:
rawCallData[rawCallData.cad_cdw_id.duplicated()]

cad_cdw_id  cad_event_number  general_offense_number  \
835500      1873964       14000010381               201410381   
835504      1873952       16000405795              2016405795   
835506      1873969       16000405783              2016405783   
1131303     1873957       16000405748              2016405748   
1131304     1873958       16000405734              2016405734   
1133136     1873928       16000405451              2016405451   
1133166     1873926       16000405578              2016405578   
1133167     1873927       16000405572              2016405572   
1133185     1873923       16000405632              2016405632   
1133280     1873921       16000405658              2016405658   
1133281     1873922       16000405655              2016405655   
1133313     1873956       16000405754              2016405754   
1133328     1873960       16000405676              2016405676   
1133329     1873961       16000405652              2016405652   
1133330     1873962       16000405630              2016405630   
1133331     1873963       16000405625              2016405625   
1133339     1873949       16000405691              2016405691   
1133357     1873951       16000405621              2016405621   
1133358     1873966       16000405803              2016405803   
1133359     1873967       16000405800              2016405800   
1133375     1873959       16000405693              2016405693   
1133827     1873965       16000405584              2016405584   
1143875     1875475       16000408522              2016408522   
1144151     1875474       16000408531              2016408531   
1144189     1875479       16000408324              2016408324   
1144566     1875477       16000408360              2016408360   
1146937     1875911       16000408923              2016408923   

         event_clearance_code  \
835500                   64.0   
835504                  280.0   
835506                  470.0   
1131303                 245.0   
1131304                 470.0   
1133136                 450.0   
1133166                  64.0   
1133167                 282.0   
1133185                 430.0   
1133280                  63.0   
1133281                 460.0   
1133313                 470.0   
1133328                  41.0   
1133329                 220.0   
1133330                  71.0   
1133331                  40.0   
1133339                 100.0   
1133357                 430.0   
1133358                 470.0   
1133359                 280.0   
1133375                 200.0   
1133827                  52.0   
1143875                 470.0   
1144151                 470.0   
1144189                  64.0   
1144566                 373.0   
1146937                 450.0   

                               event_clearance_description zone_beat  \
835500                                            SHOPLIFT        N3   
835504                                   SUSPICIOUS PERSON        B1   
835506       PARKING VIOLATION (EXCEPT ABANDONED VEHICLES)        C3   
1131303                                 DISTURBANCE, OTHER        B2   
1131304      PARKING VIOLATION (EXCEPT ABANDONED VEHICLES)        B1   
1133136                DRIVING WHILE UNDER INFLUENCE (DUI)        U1   
1133166                                           SHOPLIFT        D1   
1133167  SUSPICIOUS CIRCUMSTANCES - BUILDING (OPEN DOOR...        D1   
1133185                            MOTOR VEHICLE COLLISION        F1   
1133280                                  THEFT - CAR PROWL        U1   
1133281                         TRAFFIC (MOVING) VIOLATION        J1   
1133313      PARKING VIOLATION (EXCEPT ABANDONED VEHICLES)        D3   
1133328                                HARASSMENT, THREATS        F2   
1133329                         CRISIS COMPLAINT - GENERAL        U3   
1133330                                         AUTO THEFT        C2   
1133331                                    ASSAULTS, OTHER        E3   
1133339                   FRAUD (INCLUD

In [34]:
rawCallData.sort_values(by=['cad_cdw_id'],inplace=True) # Sort by the cad_cdw_id

In [35]:
 rawCallData.loc[rawCallData.cad_cdw_id.duplicated(keep=False) == True] # Now we can see our duplicates

cad_cdw_id  cad_event_number  general_offense_number  \
835468      1873921       14000010389               201410389   
1133280     1873921       16000405658              2016405658   
1133281     1873922       16000405655              2016405655   
835469      1873922       14000010384               201410384   
1133185     1873923       16000405632              2016405632   
835470      1873923       14000010370               201410370   
1133166     1873926       16000405578              2016405578   
835471      1873926       14000010083               201410083   
835472      1873927       14000010066               201410066   
1133167     1873927       16000405572              2016405572   
1133136     1873928       16000405451              2016405451   
835473      1873928       14000009504                20149504   
835484      1873949       14000010412               201410412   
1133339     1873949       16000405691              2016405691   
1133357     1873951       16000405621              2016405621   
835485      1873951       14000010393               201410393   
835486      1873952       14000010387               201410387   
835504      1873952       16000405795              2016405795   
835488      1873956       14000010201               201410201   
1133313     1873956       16000405754              2016405754   
1131303     1873957       16000405748              2016405748   
835489      1873957       14000010441               201410441   
1131304     1873958       16000405734              2016405734   
835490      1873958       14000010420               201410420   
1133375     1873959       16000405693              2016405693   
835481      1873959       14000010419               201410419   
835491      1873960       14000010417               201410417   
1133328     1873960       16000405676              2016405676   
1133329     1873961       16000405652              2016405652   
835492      1873961       14000010408               201410408   
835499      1873962       14000010400               201410400   
1133330     1873962       16000405630              2016405630   
835482      1873963       14000010386               201410386   
1133331     1873963       16000405625              2016405625   
835500      1873964       14000010381               201410381   
835479      1873964       16000405622              2016405622   
835483      1873965       14000010362               201410362   
1133827     1873965       16000405584              2016405584   
835501      1873966       14000010352               201410352   
1133358     1873966       16000405803              2016405803   
1133359     1873967       16000405800              2016405800   
835487      1873967       14000010334               201410334   
835506      1873969       16000405783              2016405783   
835502      1873969       14000008999                20148999   
836126      1875474       14000012503               201412503   
1144151     1875474       16000408531              2016408531   
836127      1875475       14000012470               201412470   
1143875     1875475       16000408522              2016408522   
836128      1875477       14000012420               201412420   
1144566     1875477       16000408360              2016408360   
1144189     1875479       16000408324              2016408324   
836129      1875479       14000011581               201411581   
857606      1875911       14000012658               201412658   
1146937     1875911       16000408923              2016408923   

         event_clearance_code  \
835468                  184.0   
1133280                  63.0   
1133281                 460.0   
835469                   64.0   
1133185                 430.0   
835470                  470.0   
1133166                  64.0   
835471                  363.0   
835472                  363.0   
1133167                 282.0   
1133136                 450.0   
835473                   71.0   
8354

### Well we can see here that these are definitely different events, so our ID numbers are not unique identifiers, lets check cad_event number next

In [36]:
rawCallData.cad_event_number = rawCallData.cad_event_number.astype('int64', errors='raise')

In [37]:
rawCallData.sort_values(by=['cad_event_number'], inplace=True)

In [38]:
rawCallData.loc[rawCallData.cad_event_number.duplicated(keep=False) == True]

cad_cdw_id  cad_event_number  general_offense_number  \
997757       931933       12000252943              2012252943   
1285548      579794       12000252943              2012252943   
1185456     2142696       14000275708              2014275708   
1185694     2143025       14000275708              2014275708   
1199203     2157228       14000314989              2014314989   
1200366     2158430       14000314989              2014314989   
1264603     2230530       14000422775              2014422775   
1266669     2234938       14000422775              2014422775   
1265028     2230986       14000423288              2014423288   
1266801     2235076       14000423288              2014423288   
1266558     2232734       14000425411              2014425411   
1267675     2236075       14000425411              2014425411   
1266522     2232825       14000425642              2014425642   
1266959     2235303       14000425642              2014425642   
1284814     2258678       15000025381               201525381   
1283704     2254754       15000025381               201525381   
1288534      582859       15000029809               201529809   
1284461     2257963       15000029809               201529809   
1333371     1048498       15000096986               201596986   
1334689     1050258       15000096986               201596986   
1363118     1085091       15000147931              2015147931   
1363254     1084729       15000147931              2015147931   
1373987     1097491       15000166065              2015166065   
1374182     1098250       15000166065              2015166065   
1397576     1125136       15000205924              2015205924   
1400720     1128773       15000205924              2015205924   
570218      1413020       15000229005              2015229005   
854250      1481852       15000229005              2015229005   
570217      1413019       15000229075              2015229075   
854249      1481851       15000229075              2015229075   
...             ...               ...                     ...   
851995      1480900       15000238932              2015238932   
719766      1412120       15000238932              2015238932   
851994      1480899       15000238934              2015238934   
567573      1412119       15000238934              2015238934   
719765      1412118       15000238936              2015238936   
851993      1480897       15000238936              2015238936   
719764      1412117       15000238937              2015238937   
851992      1480896       15000238937              2015238937   
851981      1480890       15000238949              2015238949   
719763      1412114       15000238949              2015238949   
851960      1480887       15000238954              2015238954   
567569      1412113       15000238954              2015238954   
862078      1517019       15000281383              2015281383   
861538      1517119       15000281383              2015281383   
930674      1582292       15000361592              2015361592   
932177      1585932       15000361592              2015361592   
967306      1648238       16000010971               201610971   
968995      1650374       16000010971               201610971   
817342      1663700       16000029094               201629094   
814594      1658699       16000029094               201629094   
158195      1664942       16000034011               201634011   
815948      1661506       16000034011               201634011   
152973      1664092       16000036928               201636928   
148428      1662608       16000036928               201636928   
224077      1673681       16000055177               201655177   
294654      1679158       16000055177               201655177   
291049      1678932       16000062288               201662288   
307914      1683497       16000062288               201662288   
679451      1733611       16000155868              2016155868   
599821      1730085       16000155868       

### So here we have duplicated records, and a valid ID column, 
Which we could just use the index and get rid of these ID columns altogether, but I would like to have a name for each record that will travel with the record as it gets moved into different dataframes as we merge in the future.
### We will drop our duplicates

In [39]:
rawCallData = rawCallData.drop_duplicates(['cad_event_number'], keep='first')

In [40]:
rawCallData.drop(['general_offense_number', 'cad_cdw_id'], axis=1, inplace=True)

In [41]:
rawCallData.event_clearance_code = rawCallData.event_clearance_code.astype('int64', errors='raise')
rawCallData.sort_values(by=['event_clearance_code'],inplace=True) 

# So now we have a nice clean dataframe that just needs some feature engineering,
We wil start with the types of events. There are 111 different codes, let's see if we can bin these into categories based on the code. First we will make a code book to see what each code means. Then we will bin the records into general types of calls.

In [42]:
eventCode = rawCallData.event_clearance_code
eventDesc = rawCallData.event_clearance_description

eventBook = dict(zip(eventCode, eventDesc))

eventBook


{10: 'HOMICIDE',
 30: 'ARMED ROBBERY',
 31: 'STRONG ARM ROBBERY',
 40: 'ASSAULTS, OTHER',
 41: 'HARASSMENT, THREATS',
 42: 'HARASSMENT, THREATS - BY TELEPHONE, WRITING',
 43: 'ASSAULTS, FIREARM INVOLVED',
 49: 'ASSAULTS, GANG RELATED',
 50: 'BURGLARY - RESIDENTIAL, UNOCCUPIED',
 51: 'BURGLARY - RESIDENTIAL, OCCUPIED',
 52: 'BURGLARY - COMMERCIAL',
 53: 'BURGLARY - UNOCCUPIED STRUCTURE ON RESIDENTIAL PROPERTY',
 61: 'THEFT - AUTO ACCESSORIES',
 62: 'BICYCLE THEFT',
 63: 'THEFT - CAR PROWL',
 64: 'SHOPLIFT',
 65: 'THEFT - MISCELLANEOUS',
 71: 'AUTO THEFT',
 72: 'AUTO THEFT AND RECOVERY',
 73: 'AUTO RECOVERY',
 74: 'LICENSE PLATE THEFT OR LOSS',
 92: 'RECKLESS BURNING',
 100: 'FRAUD (INCLUDING IDENTITY THEFT)',
 101: 'FORGERY, BAD CHECKS',
 121: 'GAMBLING',
 122: 'LIQUOR VIOLATIONS (BUSINESS)',
 124: 'PORNOGRAPHY',
 125: 'PROSTITUTION',
 126: 'VICE, OTHER',
 127: 'SOAP (STAY OUT OF AREA OF PROSTITUTION) ORDER VIOLATION',
 130: 'PROPERTY DESTRUCTION',
 139: 'GANG GRAFFITI',
 141: 'LEWD CON

### The codes seem to follow a pattern of general types of crime being grouped into a ten base system. 
The system skips around a little bit and we want to hold out some of the codes as individual bins, so we can't just linspace to generate our bins. We will have to manually create a list of our bins unfortunately, then do the same for our labels. We can then add our bins and labels to the rawCallData. Then we can tackle our other categorical variable, zone_beat

In [43]:
bins = [0,10,39,49,59,69,79,92,109,129,139,149,169,171,177,179,189,199,209,229,259,269,289,299,323,339,346,
        347,359,369,379,389,449,450,495,620]
categories = ['homicide', 'robbery', 'assault', 'burglary', 'theft', 'car_theft', 'reckless_burning', 'fraud', 
          'vice', 'property', 'lewd_behavior', 'trespass', 'park_violation', 'liquor_violation', 'drive_by_shooting', 
          'narcotics', 'warrant_service', 'security_alarms', 'mental_health', 'disturbance', 'animals', 
          'suspicious_case', 'person_with_weapon', 'harbor_theft','casualty', 'harbor_boat_assist', 'harbor_dui', 
          'hazards', 'missing_person', 'missing_property', 'crowd_control', 'traffic_general', 'dui', 
          'traffic_minor', 'human_trafficking']
rawCallData['event_type'] = pd.cut(rawCallData['event_clearance_code'], bins, labels=categories)
rawCallData['bins'] = pd.cut(rawCallData['event_clearance_code'], bins) # Add two columns with our bins and labels

In [44]:
pd.value_counts(rawCallData['event_type']) # Take a look at our totals by bin

disturbance           152722
traffic_minor         128685
suspicious_case       126012
theft                  83536
traffic_general        63670
liquor_violation       46224
security_alarms        41753
trespass               33340
burglary               26335
assault                25241
car_theft              21647
mental_health          16839
narcotics              15903
property               14435
hazards                13861
dui                    11030
missing_property       11015
fraud                  10713
warrant_service         6518
robbery                 5118
casualty                4657
park_violation          4465
missing_person          3945
lewd_behavior           2743
animals                 2394
person_with_weapon      2278
vice                    1610
harbor_boat_assist       733
drive_by_shooting        381
crowd_control            374
reckless_burning         157
homicide                  93
harbor_theft              45
harbor_dui                 5
human_traffick

### So, we have managed to cut down our 111 codes into 35 categories and binned the data as such.
We will now use the event type column to create our dummy columns and do the same with zone_beat.

In [45]:
rawCallData = pd.concat([rawCallData, pd.get_dummies(rawCallData['event_type'])], axis=1)

In [46]:
rawCallData = pd.concat([rawCallData, pd.get_dummies(rawCallData['zone_beat'])], axis=1)
rawCallData.columns = rawCallData.columns.str.lower() # Fix our column names 

In [47]:
print(rawCallData.shape)
rawCallData.head()

(878480, 133)


cad_event_number  event_clearance_code event_clearance_description  \
1079591       17000056904                    10                    HOMICIDE   
196931        16000050525                    10                    HOMICIDE   
1253712       14000405538                    10                    HOMICIDE   
525375        17000073072                    10                    HOMICIDE   
1359327       17000298506                    10                    HOMICIDE   

        zone_beat   longitude   latitude      formatted_time event_type  \
1079591        K3 -122.325060  47.597930 2017-02-15 21:15:45   homicide   
196931         D1 -122.345860  47.612990 2016-02-11 13:21:02   homicide   
1253712        O3 -122.319282  47.549830 2014-12-07 14:00:00   homicide   
525375         B3 -122.338860  47.661385 2017-02-28 17:10:27   homicide   
1359327        G3 -122.311530  47.595870 2017-08-13 01:56:59   homicide   

            bins  homicide ...  u1  u2  u3  us  w  w1  w2  w3  wp  ws  
1079591  (0, 10]         1 ...   0   0   0   0  0   0   0   0   0   0  
196931   (0, 10]         1 ...   0   0   0   0  0   0   0   0   0   0  
1253712  (0, 10]         1 ...   0   0   0   0  0   0   0   0   0   0  
525375   (0, 10]         1 ...   0   0   0   0  0   0   0   0   0   0  
1359327  (0, 10]         1 ...   0   0   0   0  0   0   0   0   0   0  

[5 rows x 133 columns]

# Looks like we finally have a nicely cleaned data frame, although it has some extra columns from our binning and making dummy columns.
We will go ahead and save this data frame to disk and create a codebook in markdown so anyone else who wants to work with this data in the future (most likely us) will have a nice and organized dataframe that can be used with dummy columns or without, binned events or not, and a nice record of what each column means. Then we can clean it up for our present use and change the name to just callData.

In [49]:
rawCallData.to_pickle('rawCallData.pkl') # We will save it as a pickle file, since we love python so much

In [53]:
callData = rawCallData.drop(['event_clearance_code', 'event_clearance_description', 
                             'zone_beat', 'event_type', 'bins'], axis=1) # Drop 'em like they're hot 

In [55]:
del rawCallData # We have great memory, the best memory, it's yuuuge, but waste not want not
callData.head()

cad_event_number   longitude   latitude      formatted_time  \
1079591       17000056904 -122.325060  47.597930 2017-02-15 21:15:45   
196931        16000050525 -122.345860  47.612990 2016-02-11 13:21:02   
1253712       14000405538 -122.319282  47.549830 2014-12-07 14:00:00   
525375        17000073072 -122.338860  47.661385 2017-02-28 17:10:27   
1359327       17000298506 -122.311530  47.595870 2017-08-13 01:56:59   

         homicide  robbery  assault  burglary  theft  car_theft ...  u1  u2  \
1079591         1        0        0         0      0          0 ...   0   0   
196931          1        0        0         0      0          0 ...   0   0   
1253712         1        0        0         0      0          0 ...   0   0   
525375          1        0        0         0      0          0 ...   0   0   
1359327         1        0        0         0      0          0 ...   0   0   

         u3  us  w  w1  w2  w3  wp  ws  
1079591   0   0  0   0   0   0   0   0  
196931    0   0  0   0   0   0   0   0  
1253712   0   0  0   0   0   0   0   0  
525375    0   0  0   0   0   0   0   0  
1359327   0   0  0   0   0   0   0   0  

[5 rows x 128 columns]

## Now we will tackle our other dataset, rawWeather

In [58]:
print(rawWeatherData.shape)
rawWeatherData.head()

(38335, 29)


dt                         dt_iso  city_id  city_name  lat  lon  \
0  1349096400  2012-10-01 13:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
1  1349186400  2012-10-02 14:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
2  1349190000  2012-10-02 15:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
3  1349193600  2012-10-02 16:00:00 +0000 UTC  5809844        NaN  NaN  NaN   
4  1349197200  2012-10-02 17:00:00 +0000 UTC  5809844        NaN  NaN  NaN   

     temp  temp_min  temp_max  pressure         ...          snow_1h  snow_3h  \
0  281.80    278.15    287.59      1027         ...              NaN      NaN   
1  281.62    278.15    286.48      1046         ...              NaN      NaN   
2  282.71    279.82    289.82      1026         ...              NaN      NaN   
3  285.05    281.48    293.15      1026         ...              NaN      NaN   
4  287.97    282.59    296.48      1027         ...              NaN      NaN   

   snow_24h  snow_today  clouds_all  weather_id  weather_main  \
0       NaN         NaN           1         800         Clear   
1       NaN         NaN          66         800         Clear   
2       NaN         NaN           1         800         Clear   
3       NaN         NaN           1         800         Clear   
4       NaN         NaN           1         800         Clear   

   weather_description  weather_icon      formatted_time  
0         sky is clear           01n 2012-10-01 06:00:00  
1         sky is clear           02d 2012-10-02 07:00:00  
2         sky is clear           01d 2012-10-02 08:00:00  
3         sky is clear           01d 2012-10-02 09:00:00  
4         sky is clear           01d 2012-10-02 10:00:00  

[5 rows x 29 columns]

### We know for starters that we have our time column squared away already, so we can lose the old columns

In [60]:
rawWeatherData = rawWeatherData.drop(['dt', 'dt_iso'], axis=1)

### It appears that our temperature columns are in Kelvin,
so we will need to convert those. We only need the temp, we don't need the range for each hour, so let's drop min and max while we're at it

In [62]:
rawWeatherData = rawWeatherData.drop(['temp_min', 'temp_max'], axis=1)

In [68]:
rawWeatherData.temp = rawWeatherData.temp.apply(lambda x: (x*(9/5))-459.67)

In [69]:
rawWeatherData.temp.head()

0    47.570
1    47.246
2    49.208
3    53.420
4    58.676
Name: temp, dtype: float64

### That is oddly specific, let's coerce to int64

In [70]:
rawWeatherData.temp = rawWeatherData.temp.astype('int64', errors='raise')

### We also know that we are only working with data from one city, Seattle, and we will use this data for the whole city, so we can lose the geo information as well

In [72]:
rawWeatherData = rawWeatherData.drop(['city_id', 'city_name', 'lat', 'lon', 'sea_level', 'grnd_level'], axis=1)

In [73]:
rawWeatherData.head()

temp  pressure  humidity  wind_speed  wind_deg  rain_1h  rain_3h  rain_24h  \
0    47      1027        81           0         0      NaN      NaN       NaN   
1    47      1046        67           1       180      NaN      NaN       NaN   
2    49      1026        87           2       140      NaN      NaN       NaN   
3    53      1026        76           0         0      NaN      NaN       NaN   
4    58      1027        58           2        30      NaN      NaN       NaN   

   rain_today  snow_1h  snow_3h  snow_24h  snow_today  clouds_all  weather_id  \
0         NaN      NaN      NaN       NaN         NaN           1         800   
1         NaN      NaN      NaN       NaN         NaN          66         800   
2         NaN      NaN      NaN       NaN         NaN           1         800   
3         NaN      NaN      NaN       NaN         NaN           1         800   
4         NaN      NaN      NaN       NaN         NaN           1         800   

  weather_main weather_description weather_icon      formatted_time  
0        Clear        sky is clear          01n 2012-10-01 06:00:00  
1        Clear        sky is clear          02d 2012-10-02 07:00:00  
2        Clear        sky is clear          01d 2012-10-02 08:00:00  
3        Clear        sky is clear          01d 2012-10-02 09:00:00  
4        Clear        sky is clear          01d 2012-10-02 10:00:00